pkg/util/kubernetes/apiserver/apiserver.go

```diff
@@ -92,10 +96,81 @@ func (c *APIClient) connect() error {
 	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
 	defer cancel()
 	version, err := c.client.Discovery().Version(ctx)
-	if err == nil {
-		log.Debugf("connected to kubernetes apiserver, version %s", version.GitVersion)
+	if err != nil {
+		return err
 	}
-	return err
+	log.Debugf("connected to kubernetes apiserver, version %s", version.GitVersion)
+	useServiceMapper := config.Datadog.GetBool("use_service_mapper")
+	if !useServiceMapper {
+		return nil
+	}
+	c.startServiceMapping()
+	return nil
+}
+
+// ServiceMapperBundle maps the podNames to the serviceNames they are associated with.
+// It is updated by mapServices in services.go
+type ServiceMapperBundle struct {
+	PodNameToServices map[string][]string
+	m                 sync.RWMutex
+}
+
+func newServiceMapperBundle() *ServiceMapperBundle {
+	return &ServiceMapperBundle{
+		PodNameToServices: make(map[string][]string),
+	}
+}
+
+// startServiceMapping is only called once, when we have confirmed we could correctly connect to the API server.
+// The logic here is solely to retrieve Nodes, Pods and Endpoints. The processing part is in mapServices.
+func (c *APIClient) startServiceMapping() {
+	tickerSvcProcess := time.NewTicker(servicesPollIntl)
+	go func() {
+		for {
+			select {
+			case <-tickerSvcProcess.C:
+				// The timeout for the context is the same as the poll frequency.
+				// We use a new context at each run, to recover if we can't access the API server temporarily.
+				// A poll run should take less than the poll frequency.
+				ctx, cancel := context.WithTimeout(context.Background(), servicesPollIntl)
+				defer cancel()
+
+				// We fetch nodes to reliably use nodename as key in the cache. Avoiding to retrieve them from the endpoints/pods.
+				nodes, err := c.client.CoreV1().ListNodes(ctx)
+				if err != nil {
+					log.Errorf("could not collect nodes from the API Server: %q", err.Error())
```

Comment:
s/could/Could

Reply:
hum I was told to always use lowercases for errors ?

Reply:
The lowercases are for the message the `error`.
```go
fmt.Errorf("lowercase")
```
But in our case, it's a log message with a severity error.